Copy button in code blocks

avinder-red-crackle opened the pull request.

thtmnisamnstr commented:
Preview URL: https://646ef3a330dd6c7bf35cc613--ephemeral-narwhal-3ac3b2.netlify.app

thtmnisamnstr commented:
@adamgordonbell We could use your feedback on this. I know it isn't 100% correct, but I'm curious how you'd like it to work. Check out this post for a good example of the issues, https://646ef3a330dd6c7bf35cc613--ephemeral-narwhal-3ac3b2.netlify.app/blog/docker-networking/.
* It doesn't show copy on all code blocks. I think it's triggered by having a file/terminal name associated with the code block.
* Some code blocks are actually terminal output and no copy button should be there.

adamgordonbell commented:
Yes, there are different types of code blocks, so I guess some decisions have to be made. 

This shows potential but this might require a couple and forth. 
This page is interesting because I think it has all the types of code blocks:
https://646ef3a330dd6c7bf35cc613--ephemeral-narwhal-3ac3b2.netlify.app/blog/awk-examples/

## Terminal Stuff
For terminal blocks, with ">_", since much of our output has prompts or output mixed in, we show the header, but not the copy button.

For those, it would be great to make it look similar to a terminal icon, with >_ in the header in green on far left and "Terminal" in the center as it is now.

## No File Name / Results
For blocks without any filename or that are results, we should not show a header or copy button. This means eliminating the double headers like this:

<img width="642" alt="Screen Shot 2023-05-25 at 4 08 11 PM" src="https://github.com/earthly/website/assets/3921401/9f9cb5de-74a8-4f89-8760-2316bcb8c932">


And also ones like this:
<img width="641" alt="Screen Shot 2023-05-25 at 4 18 43 PM" src="https://github.com/earthly/website/assets/3921401/617842fc-4705-470d-88a0-9bbb406011f0">

## Have file name

The blocks with a filename ( that isn't ">_") look good I think!
<img width="761" alt="Screen Shot 2023-05-25 at 4 09 26 PM" 
src="https://github.com/earthly/website/assets/3921401/478a89c0-faaa-4f62-887b-c8a79f9f9bea">



## Optional Copy

It would be nice if we (Josh and I ) could go back and make the terminal code copyable and then turn on the copy buttons on those one by one.  Can we have a flag in markdown to do this?

So if block contains config contains 'copy' we show a copy button regardless of the fact it is a terminal?
 ```
~~~{.bash caption=">_"  copy}
    git clone https://github.com/earthly/example-grpc-key-value-store.git
    cd example-grpc-key-value-store/go-server
    earth +kvserver-docker
~~~
```

The 'copy' would bubble up into HTML as the class and so should work similar to how `caption=bla` works but we should test it by using it on some page. 

That is my thinking.

@jalletto  or anyone else have thoughts on this?

adamgordonbell commented:
Hi @avinder-red-crackle ,
Getting better. This one looks strange:

https://64758830cecdad2291d48d1b--ephemeral-narwhal-3ac3b2.netlify.app/blog/awk-examples/

![Screen Shot 2023-05-30 at 9 05 42 AM](https://github.com/earthly/website/assets/3921401/61dd4c0f-28a7-4869-b019-8aa01215dea0)

avinder-red-crackle commented:
@adamgordonbell That is an old link. Check this: https://64760b3577ded30073f33d65--ephemeral-narwhal-3ac3b2.netlify.app/blog/awk-examples/

jalletto commented:
> @adamgordonbell That is an old link. Check this: https://64760b3577ded30073f33d65--ephemeral-narwhal-3ac3b2.netlify.app/blog/awk-examples/

This looks good to me.

adamgordonbell commented:
Ok, one last change @avinder-red-crackle . Can we just show this header and copy button for blocks that have a filename (  that isn't ">_". Everything else we will leave without them for now. 

In retrospect I think the header is too big for the little snippets but works well for the larger code blocks with filenames.

avinder-red-crackle commented:
Preview URL: https://64799f3d2d520a6741a2641a--ephemeral-narwhal-3ac3b2.netlify.app

@adamgordonbell Please review

adamgordonbell reviewed (changes requested):
see https://github.com/earthly/website/pull/501#issuecomment-1573499452

avinder-red-crackle commented:
@adamgordonbell All the addressed issues are fixed. What should be shown for code blocks with  bcaption? A header with copy button, if it is not ">_", "Output" & "Result". Else green text in right bottom corner?

avinder-red-crackle commented:
> > @adamgordonbell All the addressed issues are fixed. What should be shown for code blocks with bcaption? A header with copy button, if it is not ">_", "Output" & "Result". Else green text in right bottom corner?
> 
> Yes, exactly!

I've made the changes already. Please check here: https://647ed7938817b470ae78062d--ephemeral-narwhal-3ac3b2.netlify.app

thtmnisamnstr commented:
@adamgordonbell Any feedback on the copy button?

adamgordonbell commented:
> I've made the changes already

Oh, I didn't realize that! I've reviewed and it looks good.
